client/blocks/qr-code-login/index.jsx

```diff
@@ -0,0 +1,266 @@
+import { getTracksAnonymousUserId } from '@automattic/calypso-analytics';
+import config from '@automattic/calypso-config';
+import { Card } from '@automattic/components';
+import { useTranslate } from 'i18n-calypso';
+import QRCode from 'qrcode.react';
+import { useEffect, useState, useMemo } from 'react';
+import qrCenter from 'calypso/assets/images/qr-login/wp.png';
+import ExternalLink from 'calypso/components/external-link';
+import { setStoredItem, getStoredItem } from 'calypso/lib/browser-storage';
+import { useInterval } from 'calypso/lib/interval';
+
+import './style.scss';
+
+const AUTH_PULL_INTERVAL = 5000; // 5 seconds
+const LOCALE_STORAGE_KEY = 'qr-login-token';
+
+const isStillValidToken = ( tokenData ) => {
+	if ( ! tokenData?.expires ) {
+		return false;
+	}
+	return tokenData.expires > Date.now() / 1000;
+};
+
+const getLoginActionResponse = async ( action, args ) => {
+	const url = `https://wordpress.com/wp-login.php?action=${ action }`;
+	const body = new URLSearchParams( {
+		...args,
+		client_id: config( 'wpcom_signup_id' ),
+		client_secret: config( 'wpcom_signup_key' ),
+	} );
+
+	const response = await fetch( url, {
+		method: 'POST',
+		credentials: 'include',
+		body,
+	} );
+
+	let json_response = null;
+	try {
+		json_response = await response.json();
+	} catch {
+		json_response = false;
+	}
+
+	return json_response;
+};
+
+const fetchQRCodeData = async ( setTokenData, tokenData, anonymousUserId, setIsErrorState ) => {
+	if ( isStillValidToken( tokenData ) ) {
+		return;
+	}
+
+	if ( ! anonymousUserId ) {
+		return;
+	}
+	// tokenData is set to null initially.
+	// Lets wait till it is set to false when the local data is the just yet.
+	if ( tokenData === null ) {
+		return;
+	}
+
+	const responseData = await getLoginActionResponse( 'qr-code-token-request-endpoint', {
+		anon_id: anonymousUserId,
+	} );
+
+	if ( responseData === false ) {
+		setIsErrorState( true );
+		return;
+	}
+
+	if ( isStillValidToken( responseData.data ) ) {
+		setTokenData( responseData.data );
+		setStoredItem( LOCALE_STORAGE_KEY, responseData.data );
+		return;
+	}
+	setIsErrorState( true );
+	return;
+};
+
+const fetchAuthState = async (
```

Comment:
I would consider moving both `fetch*` function _inside_ the `TokenQRCode` component. That would remove all parameters, as the variables are in the closure.

The fact that `fetchAuthState` has 7 parameters, somewhat random ones, suggests that it's not really an independent function.

Reply:
Good call.